### PR TITLE
Added Code for Lift Subsystem

### PR DESCRIPTION
**Pull request by avakimc**
- replaced Scorer subsystem with Lift subsystem
- wrote methods to deploy and retract elevator, unfold and fold arm, and set both to given position based on TreeMap
- made placeholder variables in constants and calibrations

### CODE REVIEW

File: src/main/java/frc/robot/Calibrations.java [hunk 1]
```diff
@@ -33,4 +33,20 @@ public final class Calibrations {
...
   public static final double INTAKE_INTAKING_POWER = 1.0;
   public static final double INTAKE_EJECTION_POWER = -1.0;
+
+  public static final double ELEVATOR_DEPLOYING_POWER = 1.0;
+  public static final double ELEVATOR_RETRACTING_POWER = -1.0;
+
+  public static final double ARM_UNFOLDING_POWER = 1.0;
+  public static final double ARM_FOLDING_POWER = -1.0;
+
+  /** Input rotations, output [0,1] */
```
Comment (balghane) [inline src/main/java/frc/robot/Calibrations.java]: Let's change this to input inches

File: src/main/java/frc/robot/Constants.java [hunk 2]
```diff
@@ -85,4 +85,14 @@ public static final class SwerveDriveConstants {
     
     public static final boolean GYRO_REVERSED = false;
   }
+
+  /* Scalar for elevator motor encoder from RPM to real degrees per seconds */
+  public static final double ELEVATOR_MOTOR_ENCODER_VELOCITY_SCALAR = PLACEHOLDER_DOUBLE;
+  /* Scalar for elevator motor encoder from rotations to real degrees */
+  public static final double ELEVATOR_MOTOR_ENCODER_SCALAR = PLACEHOLDER_DOUBLE;
```
Comment (balghane) [inline src/main/java/frc/robot/Constants.java]: Inches rather than degrees

File: src/main/java/frc/robot/RobotMap.java [hunk 3]
```diff
@@ -26,6 +26,10 @@ public final class RobotMap {
   /** Intake Spark MAX CAN IDs */
   public static final int INTAKE_LEFT_MOTOR_CAN_ID = 22,
     INTAKE_RIGHT_MOTOR_CAN_ID = 23;
+
+  /** Lift Spark MAX CAN IDs */
+  public static final int LIFT_MOTOR_CAN_ID = 37,
```
Comment (balghane) [inline src/main/java/frc/robot/RobotMap.java]: elevator rather than lift here

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+  TreeMap<LiftPositions, Pair<Double, Double>> liftPositionsMap;
+
+
+  /** Creates a new Lift */
+  public Lift() {
+    elevator.restoreFactoryDefaults();
+    arm.restoreFactoryDefaults();
+
+    /* Get positions and degrees of elevator through encoder*/
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/Lift.java]: Inches

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    liftPositionsMap = new TreeMap<LiftPositions, Pair<Double, Double>>();
+    liftPositionsMap.put(LiftPositions.GRAB_FROM_INTAKE, new Pair<Double, Double>(Calibrations.PLACEHOLDER_DOUBLE, Calibrations.PLACEHOLDER_DOUBLE));
+    liftPositionsMap.put(LiftPositions.SHELF, new Pair<Double, Double>(Calibrations.PLACEHOLDER_DOUBLE, Calibrations.PLACEHOLDER_DOUBLE));
+    liftPositionsMap.put(LiftPositions.SCORE_MID, new Pair<Double, Double>(Calibrations.PLACEHOLDER_DOUBLE, Calibrations.PLACEHOLDER_DOUBLE));
+    liftPositionsMap.put(LiftPositions.SCORE_HIGH, new Pair<Double, Double>(Calibrations.PLACEHOLDER_DOUBLE, Calibrations.PLACEHOLDER_DOUBLE));
+    liftPositionsMap.put(LiftPositions.STOWED, new Pair<Double, Double>(Calibrations.PLACEHOLDER_DOUBLE, Calibrations.PLACEHOLDER_DOUBLE));
+  }
+
+
+  public enum LiftPositions {
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/Lift.java]: This should be singular not plural. If you have an enum value, it indicates one position

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    GRAB_FROM_INTAKE,
+    SHELF,
+    SCORE_MID,
+    SCORE_HIGH,
+    STOWED
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/Lift.java]: Let's rename STOWED to STARTING

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+  /** Pushes the lift down */
+  public void retractElevator() {
+    elevatorPID.setReference(Calibrations.ELEVATOR_RETRACTING_POWER, CANSparkMax.ControlType.kVelocity);
+  }
+
+  /** Unfolds the arm */
+  public void unfoldArm() {
+    armPID.setReference(Calibrations.ARM_UNFOLDING_POWER, CANSparkMax.ControlType.kVelocity);
+  }
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/Lift.java]: We can probably just remove these four functions

File: src/main/java/frc/robot/Calibrations.java [hunk 8]
```diff
@@ -33,4 +33,20 @@ public final class Calibrations {
 
   public static final double INTAKE_INTAKING_POWER = 1.0;
   public static final double INTAKE_EJECTION_POWER = -1.0;
+
+  public static final double ELEVATOR_DEPLOYING_POWER = 1.0;
+  public static final double ELEVATOR_RETRACTING_POWER = -1.0;
+
+  public static final double ARM_UNFOLDING_POWER = 1.0;
+  public static final double ARM_FOLDING_POWER = -1.0;
```
Comment (balghane) [inline src/main/java/frc/robot/Calibrations.java]: We probably don't need these anymore

**Review by balghane — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+import java.util.TreeMap;
+
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.RelativeEncoder;
+import com.revrobotics.SparkMaxPIDController;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+
+
+public class Lift extends SubsystemBase {
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/Lift.java:21]: Can we add a comment for the whole lift that says it contains the elevator, arm, and game piece grabber

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+  private final RelativeEncoder elevatorEncoder;
+  private final RelativeEncoder armEncoder;
+
+  private SparkMaxPIDController elevatorPID =  elevator.getPIDController();
+
+  private CANSparkMax arm = new CANSparkMax(RobotMap.ARM_MOTOR_CAN_ID, MotorType.kBrushless);
+  private SparkMaxPIDController armPID =  arm.getPIDController();
+
+  TreeMap<LiftPosition, Pair<Double, Double>> liftPositionMap;
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/Lift.java:36]: This needs a comment to say what this Map is. It indicates the elevator and arm positions associated with each position of the lift. We should be super explicit about what the first value and second value refer to

**Review by balghane — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Calibrations.java [hunk 11]
```diff
@@ -33,4 +33,14 @@ public final class Calibrations {
 
   public static final double INTAKE_INTAKING_POWER = 1.0;
   public static final double INTAKE_EJECTION_POWER = -1.0;
+
+  /** Input inches, output [0,1] */
+  public static final double ELEVATOR_P = PLACEHOLDER_DOUBLE,
+  ELEVATOR_I = PLACEHOLDER_DOUBLE,
+  ELEVATOR_D = PLACEHOLDER_DOUBLE;
+
+  /** Input rotations, output [0,1] */
```
Comment (balghane) [inline src/main/java/frc/robot/Calibrations.java]: degrees

File: src/main/java/frc/robot/Constants.java [hunk 12]
```diff
@@ -85,4 +85,14 @@ public static final class SwerveDriveConstants {
     
     public static final boolean GYRO_REVERSED = false;
   }
+
+  /* Scalar for elevator motor encoder from RPM to real inches per seconds */
+  public static final double ELEVATOR_MOTOR_ENCODER_VELOCITY_SCALAR = PLACEHOLDER_DOUBLE;
+  /* Scalar for elevator motor encoder from rotations to real inches */
+  public static final double ELEVATOR_MOTOR_ENCODER_SCALAR = PLACEHOLDER_DOUBLE;
```
Comment (balghane) [inline src/main/java/frc/robot/Constants.java:100]: degrees

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+  private final RelativeEncoder armEncoder;
+
+  private SparkMaxPIDController elevatorPID =  elevator.getPIDController();
+
+  private CANSparkMax arm = new CANSparkMax(RobotMap.ARM_MOTOR_CAN_ID, MotorType.kBrushless);
+  private SparkMaxPIDController armPID =  arm.getPIDController();
+
+  /** Indicates the elevator and arm positions at each position of the lift.
+   * The first value indicates the elevator position
+   * and the second value indicates the arm position
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/Lift.java]: inches and degrees

**Review by balghane — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Lift.java (hunk 9 above)
Comment (avakimc) [reply src/main/java/frc/robot/subsystems/Lift.java:21]: ok

File: src/main/java/frc/robot/subsystems/Lift.java (hunk 10 above)
Comment (avakimc) [reply src/main/java/frc/robot/subsystems/Lift.java:36]: ok, I also added the units

File: src/main/java/frc/robot/Calibrations.java (hunk 11 above)
Comment (avakimc) [reply src/main/java/frc/robot/Calibrations.java]: ok

File: src/main/java/frc/robot/Constants.java (hunk 12 above)
Comment (avakimc) [reply src/main/java/frc/robot/Constants.java:100]: ok

Comment (balghane) [reply src/main/java/frc/robot/Constants.java:100]: wait no this one is fine

File: src/main/java/frc/robot/subsystems/Lift.java (hunk 13 above)
Comment (avakimc) [reply src/main/java/frc/robot/subsystems/Lift.java]: ok

**Review by balghane — APPROVED** [review]

**Pull request merged**